### ISSUE TITLE
吕文辉任务1,2提交

### ISSUE THREAD

**Issue by WrenLv** [slwang-ustc/USTC_INT_INFER#2]
完成基本任务要求，代码位于[whlv](https://github.com/slwang-ustc/USTC_INT_INFER/tree/whlv)分支

**Comment by Tomorrowdawn** [comment]
tokenizer没考虑pad问题.
 
(跨文件导入的时候最好用一个别名标识(文件路径找到文件夹的上一层即可), batchify那里突然冒出一个utils但是找不到文件很奇怪)

该更新分支了

**Comment by WrenLv** [comment]
# 任务三和四

## sampling

- **top-k**

  sample时选择概率最高的k个token，将其他token概率设置为0，归一化，采样

- **top-p**

  不固定选择top k个token，而是从高到低依次选择，使累积概率超过p，然后在选择token中采样

- **greedy**

  选择概率最高的token（argmax）

- **temperature**

  对logits进行应用temperature：$logits = logits / temperature$ 后再进行softmax，控制采样的随机性。
  
- **beam search**

  beam search 保持多个候选beam，并在每一步选择概率最高的 beam size 个扩展序列。

![beam](https://github.com/user-attachments/assets/5331f451-ccbc-426c-9260-114235adede6)


只实现batch size为1时，可将不同beam组织成一个batch size为 beam size的批次……

```python
prompt_ids = [[1,2,3]]
logits = model(prompt_ids)
expand_prompt_ids = beam_search(logits)
# maybe expand_prompt_ids = [[1,2,3,4], [1,2,3,5]]#beam size 为2
logits = model(expand_prompt_ids)
expand_prompt_ids = beam_search(logits)#此处logits批数为2，从2*beam size中选择累计prob最高的beams
```

kvcache有待实现……可能会出现需要合并的情况，如第二轮选择 do， is

## RoPE(Rotary Position Embedding)

通过对二维向量进行旋转，再对旋转后的向量点乘，其结果一定会包含**旋转弧度的相对信息**，若$q_1$，$k_1$旋转$\theta_1$角，$k_2$旋转$\theta_2$角。则$q_1^T\cdot k_1$与$q_1^T\cdot k_2$会体现出旋转角的区别，从而体现出相对位置的区别。假设第m个token旋转m角度，第n个token旋转n角度，则 n-m 体现出两个token的相对位置信息。

$$(R_m \cdot q_m)^T\cdot(R_n \cdot k_n) = (\begin{bmatrix}\cos m&-\sin m\\\sin m&\cos m\end{bmatrix}\cdot\begin{bmatrix}q_0\\\\\\\\\\\\ q_1\end{bmatrix})^T\cdot (\begin{bmatrix}\cos n&-\sin n\\\sin n&\cos n\end{bmatrix}\cdot\begin{bmatrix}k_0\\\\\\\\\\\\k_1\end{bmatrix})$$

当维度大于2时，可以两两组合编码。

$$R_m = \begin{bmatrix}\cos m\theta_0&-\sin m\theta_0 &0&0&\cdots&\cdots&0&0\\\\\\\\\\\\\sin m\theta_0&\cos m\theta_0 &0&0&\cdots&\cdots&0&0\\\\\\\\\\\\0&0&\cos m\theta_1&-\sin m\theta_1&\cdots&\cdots&0&0\\\\\\\\\\\\0&0&\sin m\theta_1&\cos m\theta_1&\cdots&\cdots&0&0\\\\\\\\\\\\\vdots&\vdots&\vdots&\vdots&\ddots&\ddots&\vdots&\vdots\\\\\\\\\\\\\vdots&\vdots&\vdots&\vdots&\ddots&\ddots&\vdots&\vdots\\\\\\\\\\\\ 0&0&0&0&\cdots&\cdots&\cos m\theta_{d/2-1}&-\sin m\theta_{d/2-1}\\\\\\\\\\\\0&0&0&0&\cdots&\cdots&\sin m\theta_{d/2-1}&\cos m\theta_{d/2-1} \end{bmatrix}$$

$$R_m \cdot q_m = \begin{bmatrix} q0\\\\\\\\\\\\q1\\\\\\\\\\\\\vdots\\\\\\\\\\\\q_{d-2}\\\\\\\\\\\\q_{d-1} \end{bmatrix}\otimes \begin{bmatrix} \cos m\theta_0\\\\\\\\\\\\\cos m\theta_0\\\\\\\\\\\\\vdots\\\\\\\\\\\\cos m\theta_{d/2-1}\\\\\\\\\\\\cos m\theta_{d/2-1} \end{bmatrix} + \begin{bmatrix} q1\\\\\\\\\\\\q0\\\\\\\\\\\\\vdots\\\\\\\\\\\\q_{d-1}\\\\\\\\\\\\q_{d-2} \end{bmatrix}\otimes \begin{bmatrix} -\sin m\theta_0\\\\\\\\\\\\\sin m\theta_0\\\\\\\\\\\\\vdots\\\\\\\\\\\\-\sin m\theta_{d/2-1}\\\\\\\\\\\\\sin m\theta_{d/2-1} \end{bmatrix}$$

hf的实现中不是 i 与 i+1组合，而是 i 与 i+d/2组合（`rotary_half()`）

$$
R_m \cdot q_m = \begin{bmatrix} q0&q1 \\\\\\\\\\\\ \vdots \\\\\\\\\\\\ q_{d-2} \\\\\\\\\\\\ q_{d-1} \end{bmatrix}\otimes \begin{bmatrix} \cos m\theta_0 \\\\\\\\\\\\ \cos m\theta_0 \\\\\\\\\\\\ \vdots \\\\\\\\\\\\ cos m\theta_{d/2-1} \\\\\\\\\\\\ cos m\theta_{d/2-1} \end{bmatrix} + \begin{bmatrix} -q_{d/2} \\\\\\\\\\\\ -q_{d/2+1} \\\\\\\\\\\\ \vdots \\\\\\\\\\\\ q_{d/2-2} \\\\\\\\\\\\ q_{d/2-1} \end{bmatrix}\otimes \begin{bmatrix} \sin m\theta_0 \\\\\\\\\\\\ \sin m\theta_0 \\\\\\\\\\\\ \vdots \\\\\\\\\\\\ \sin m\theta_{d/2-1} \\\\\\\\\\\\ \sin m\theta_{d/2-1} \end{bmatrix}
$$

对原有RoPE的修改

1. 并非每次计算对应的position ids 而是计算全序列的 position ids，shape [bs, sequence_length]
2. 正常计算cos, sin， shape [bs, sequence_length, d]
3. 在kvcache.update()阶段进行 `apply_rotary_pos_emb`
   - 若为DynamicCache，先对q, k 编码后，存储到cache，这时取cos[bs, -1:, d]
   - 若为SinkCache，将该token存入cache，取出时，对q, k 分别编码，一般q取cos[bs, -1:, d]，k取所有cos，完毕后返回qkv state

**Comment by WrenLv** [comment]
以上对HF transformer的apply rotary过程描述有误，rotary_half()后矩阵相乘（施加旋转编码）应为：
![屏幕截图 2024-10-16 001821](https://github.com/user-attachments/assets/909c9796-f588-40bb-b023-1107eb7c852e)
cos和sin矩阵上半部分和下半部分都为：
$\cos m\theta_0$ ~ $\cos m\theta_{d/2-1}$
从而将i 与 i + d/2对应

**Comment by Tomorrowdawn** [comment]
还可以。（tips:你可以直接edit而不需要额外发一条comment. 不然comment数容易爆炸。）
顺带问问，你觉得给定两个输入A， B
1. 把它们拼成[A,B]送进LLM
2. 把它们分别送进LLM后产生的KV Cache拼成[kv_A, kv_B]送进LLM

会是等价的吗？为什么？

**Comment by WrenLv** [comment]
相当于把序列长度增加是吗？单纯地把B添加到A的后面。这样的话使用SinkCache应该是等价的，存入cache的k没有旋转编码，两种情况下Cache相同。如果使用DynamicCache(先RoPE，再存储)，是不等价的，比如kv_B的第一个token旋转角度是0， 而拼成[A,B]送入LLM的情况下，对应B第一个token旋转角度是 length(A) * \theta

**Comment by Tomorrowdawn** [comment]
你可以写个脚本实验一下。
第一趟用a，保存kv cache
第二趟用b，保存kv cache
第三趟用ab，保存kv cache
然后看看前两者拼起来是否等于后者

**Comment by WrenLv** [comment]
若使用PreRopeCache，则所有层 kvab != [kva,kvb]
若使用PostRopeCache, 假设三趟输入分别为X_a X_b X_ab
则第一层的输入满足 X_a = X_ab[0:len(a),:]， X_b = X_ab[len(a):,:]因此对于第一层的cache，两者相等。
然而第一层计算attention时会产生不同的输出结果（ab中b会计算对于a的分数），因此不满足 X_b = X_ab[len(a):,:]，第二层及以后的cache则不相等，且差异不断累积
PreRopeCache：
![屏幕截图 2024-10-16 190734](https://github.com/user-attachments/assets/e51ccdeb-8356-4fa1-b764-e890c78c78c4)
PostRopeCache：
![屏幕截图 2024-10-16 190823](https://github.com/user-attachments/assets/67660c27-d1ca-4c99-b669-54f0dbfabaaa)

**Comment by Tomorrowdawn** [comment]
ok. 这个误区要小心. 
必要任务部分应该已经结束, 不过维护一个好用的llama codebase很重要(你的任何idea想要快速验证都需要用到它), 很推荐用软工实践多改造一下, 让它更顺手.
